keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,576 @@
+
+There could be conditions that cannot be reported by a CSI driver. There could be network failure where Kubernetes may not be able to get response from the CSI driver. In this case, a call to the CSI driver may time out. There could be network congestion which causes slow response. One or more nodes where the volume is attached to may be down. This can be monitored and detected by the volume health controller so that user knows what has happened.
+
+The Kubernetes components that monitor the volumes and report events with volume health information include the following:
+
+* An external monitoring agent on each kubernetes worker node.
+* An external monitoring controller on the master node.
+
+Details will be described in the following proposal section.
+
+## Proposal
+
+Volume monitoring is the main focus of this proposal. Reactions are not in the scope of this proposal.
+
+- If a volume provisioned by the CSI driver is deleted, we need to report an event and log a message to inform users.
+
+- If mounting error occurs, we need to report an event and log a message.
+
+- If other errors occur, we also need to report an event and log a message.
+
+The main architecture is shown below:
+
+![pv health monitor architecture](./pv-health-monitor.png)
+
+The following areas will be the focus of this proposal at first:
+
+- The health condition checking of volumes themselves. For example, whether the volume is deleted, whether the usage is reaching the threshold, and so on.
+- Mounting conditions checking.
+- Other errors that could affect the usability of the volume.
+
+Three main parts are involved here in the architecture.
+
+- External Controller:
+  - The external controller will be deployed as a sidecar together with the CSI controller driver, similar to how the external-provisioner sidecar is deployed.
+  - Trigger controller RPC to check the health condition of the CSI volumes.
+  - The external controller sidecar will also watch for node failure events.
+
+- External Agent:
+  - The external agent will be deployed as a sidecar together with the CSI node driver on every Kubernetes worker node.
+  - Trigger node RPC to check volume's mounting conditions.
+  - Note that currently we do not have CSI support for local storage. As a workaround, we may check the local volumes directly by the agent at first, and then move the checks to RPC interfaces when it is ready.
```

Comment:
Is this actually going to be part of the plan? Where would this go?

Reply:
Nick said he will do this, but there's no detailed design on how to do it.  I'm going to remove this for now.  It can always be added later when we have a concrete plan.